src/psyclone/psyir/nodes/omp_directives.py

```diff
@@ -227,6 +237,814 @@ class OMPSerialDirective(OMPRegionDirective, metaclass=abc.ABCMeta):
+                        satisfiable = satisfiable and self._compare_ranges(
+                                index, array2.indices[i])
+                    else:
+                        # The only remaining option is that the indices are
+                        # References or BinaryOperations
+                        satisfiable = satisfiable and self._compare_ref_binop(
+                                index, array2.indices[i], task1, task2)
+
+            # If we have an unsatisfiable dependency between two tasks, then we
+            # need to have a taskwait between them always. We need to loop up
+            # to find these tasks' parents that is closest to the schedule
+            # which contains both tasks, and use that as the nodes which are
+            # dependent.
+            if not satisfiable:
+                # Find the lowest schedule containing both nodes.
+                schedule1 = task1.ancestor(Schedule)
+                schedule2 = task2.ancestor(Schedule)
+                # While they're not distinct
+                while schedule1 is not schedule2:
+                    if schedule1.depth > schedule2.depth:
+                        schedule1 = schedule1.ancestor(Schedule)
+                    else:
+                        schedule2 = schedule2.ancestor(Schedule)
+                task1_proxy = task1
+                while task1_proxy.parent is not schedule1:
+                    task1_proxy = task1_proxy.parent
+                task2_proxy = task2
+                while task2_proxy.parent is not schedule1:
+                    task2_proxy = task2_proxy.parent
+
+                # Now we have the closest nodes to the closest ancestor
+                # schedule, so add them to the dependent_nodes list.
+                if task1_proxy is not task2_proxy:
+                    # If they end up with the same proxy, they have the same
+                    # ancestor tree but are in different schedules. This means
+                    # that they are in something like an if/else block with
+                    # one node in an if block and the other in the else block.
+                    # These dependencies we can ignore as they are not ever
+                    # both executed
+                    dependent_nodes.append((task1_proxy, task2_proxy))
+                    lowest_position_nodes.append(min(task1_proxy.abs_position,
+                                                     task2_proxy.abs_position))
+                    highest_position_nodes.append(
+                            max(task1_proxy.abs_position,
+                                task2_proxy.abs_position))
+
+        # Need to sort lists by highest_position_nodes value, and then
+        # by lowest value if tied.
+        # Based upon
+        # https://stackoverflow.com/questions/9764298/how-to-sort-two-
+        # lists-which-reference-each-other-in-the-exact-same-way
+        if len(dependent_nodes) > 0:
```

Comment:
Just so that we're not relying on stackoverflow here, please could you document what `sorted_highest_positions`, `sorted_lowest_positions` and `sorted_dependency_pairs` contain and their datatypes. I ask because I'd convinced myself that they were each lists but then I got to some code below where you index into one of them as `[0][1]`.

Reply:
I will do, I wanted to leave the source for the code to avoid any issues in the future as well.

I think documenting what the contain is important too though.